### ISSUE TITLE
[cmd/mdatagen] Add support for declaring feature gates in metadata.yaml file

### ISSUE THREAD

**Issue by mx-psi** [open-telemetry/opentelemetry-collector-contrib#21801]
### Component(s)

cmd/mdatagen

### Describe the issue you're reporting

Components can use [feature gates][1] to have support for experimental functionality. We can add these to `metadata.yaml` together with a description. This information can then be used for code generation, documentation generation and by other consumers of the `metadata.yaml` file (e.g. it could be useful for a remote management system that wants to show a list of available feature gates).

Note that some feature gates do not fit a specific component (e.g. some of the ones we have in the core distro, or [the one on `pkg/translator/prometheus`][2]). This would not be listed for now.

Follow up tasks for individual components once support for this is available:

- [ ] Move feature gate(s) to metadata.yaml for receiver/elasticsearchreceiver
- [ ] Move feature gate(s) to metadata.yaml for receiver/dockerstatsreceiver
- [ ] Move feature gate(s) to metadata.yaml for receiver/prometheusreceiver
- [ ] Move feature gate(s) to metadata.yaml for processor/servicegraphprocessor
- [ ] Move feature gate(s) to metadata.yaml for receiver/postgresqlreceiver
- [ ] Move feature gate(s) to metadata.yaml for exporter/googlecloudexporter
- [ ] Move feature gate(s) to metadata.yaml for exporter/datadogexporter
- [ ] Move feature gate(s) to metadata.yaml for processor/spanmetricsprocessor

[1]: https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate#collector-feature-gates
[2]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/translator/prometheus#full-normalization

**Comment by psk001** [comment]
Hi @mx-psi , is this issue still up for grab? Can I take it?

**Comment by mx-psi** [comment]
> Hi @mx-psi , is this issue still up for grab? Can I take it?

Sure, feel free to work on it @psk001 :) The first step is to make the necessary changes on `cmd/mdatagen` to support this, by adding the new feature gates section.

cc @mackjmr since you were also interested on this one

**Comment by psk001** [comment]
Great, starting with it

**Comment by psk001** [comment]
Needed a little help on it, I am referring to [featureGate in opentelemetry-collector](https://github.com/open-telemetry/opentelemetry-collector/blob/main/featuregate/gate.go#L10). Is this the correct way to implement it?

**Comment by mx-psi** [comment]
@psk001 yes, that's what we are talking about :) We would need to define an equivalent YAML schema to then generate the code that creates a feature gate for each of the components listed above

**Comment by psk001** [comment]
Great, on it

**Comment by psk001** [comment]
> Components can use [feature gates](https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate#collector-feature-gates) to have support for experimental functionality. We can add these to metadata.yaml together with a description.

In the metadata.yaml file , is this how it should be added ? 

```yaml
feature-gates:
  SplitTraces:
    description: |
      It returns one ptrace.
      Traces for each trace in the given ptrace.Traces input. Each of the resulting ptrace.Traces contains exactly one trace.
  
  SplitLogs:
    description: |
      It returns one plog.
      Logs for each trace in the given plog.Logs input. Each of the resulting plog.Logs contains exactly one log.```

**Comment by narcis96** [comment]
I am continuing the work on this feature
